subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/JdkIllegalReflectionTestWorkerIntegrationTest.groovy

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.workers.internal
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.DefaultTestExecutionResult
+import org.gradle.util.Requires
+import org.gradle.util.TestPrecondition
+import spock.lang.Issue
+
+import static org.hamcrest.CoreMatchers.containsString
+
+/**
+ * Ensures test behavior and actual application behavior are equivalent when
+ * production code attempts to perform reflection on JDK internals.
+ */
+class JdkIllegalReflectionTestWorkerIntegrationTest extends AbstractIntegrationSpec {
+
+    def setup() {
+        buildFile << """
+            plugins {
+                id 'application'
+            }
+
+            ${mavenCentralRepository()}
+
+            testing.suites.test.useJUnit()
+
+            application {
+                mainClass = 'example.Main'
+            }
+        """
+
+        file("src/main/java/example/Main.java") << """
+            package example;
+
+            import java.lang.invoke.MethodHandles;
+
+            public class Main {
+                public static void main(String[] args) throws Exception {
+                    MethodHandles.privateLookupIn(java.lang.ClassLoader.class, MethodHandles.lookup());
+                }
+            }
+        """
+
+        file("src/test/java/example/MainTest.java") << """
+            package example;
+
+            import org.junit.Test;
+
+            public class MainTest {
+                @Test
+                public void runTest() throws Exception {
+                    Main.main(new String[0]);
+                }
+            }
+        """
+    }
+
+    @Requires(TestPrecondition.JDK16_OR_LATER)
+    @Issue("https://github.com/gradle/gradle/issues/19771")
+    def "both tests and production code fail when application uses illegal reflection"() {
+        when:
+        executer.withStackTraceChecksDisabled()
+        fails "run"
```

Comment:
Seems they are both used pretty often. Searching in intellij for `then:\n[\s]*fails` and `when:\n[\s]*fails` shows 336 and 775 results respectively.

Reply:
That's really surprising to me, I thought we were more consistently using the first way.  I mean, It clearly works to add `succeeds/fails` whenever, since throwing an unexpected, unhandled exception fails the test from anywhere.  But I had understood the idiomatic Spock usage to be to [use the `then/expect` blocks for assertions](https://spockframework.org/spock/docs/2.0/all_in_one.html#_blocks). 

I guess you could argue that we're making the build fail as a stimulus, then asserting about how it failed.  That seems less correct to me than setting up the build as a stimulus, then asserting it fails + details about the failure.  Maybe it's just a matter of taste.

Reply:
Updated to move `fails` into `when` and `expect`